### PR TITLE
Streaming add remfile

### LINKED ISSUE #1771: [Documentation]: RemFile
### What would you like changed or added to the documentation and why?

Add support for RemFile and add RemFile to the tutorial for streaming data

https://github.com/magland/remfile

### Do you have any interest in helping write or edit the documentation?

No.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you checked the [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this change was not already [requested](https://github.com/NeurodataWithoutBorders/pynwb/issues)?

### PR DESCRIPTION
**Pull request by bendichter**
## Motivation

remfile is a new package that has some advantages for streaming NWB HDF5 files from S3, so I want to document it as an option.

Depends on https://github.com/hdmf-dev/hdmf/pull/946

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Review by rly — DISMISSED** [review]

**Comment by rly** [comment]
@bendichter can we move this out of draft mode?

File: docs/gallery/advanced_io/streaming.py [hunk 1]
```diff
@@ -122,6 +122,30 @@
...
+# Then remfile can be used similarly to fsspec
+
+import h5py
+from pynwb import NWBHDF5IO
+import remfile
+
+file = remfile.File(s3_url)
+
+with h5py.File(file, "r") as f:
+    with NWBHDF5IO(file=file, load_namespaces=True) as io:
```
Comment (bendichter) [inline docs/gallery/advanced_io/streaming.py]:
```suggestion
    with NWBHDF5IO(file=f, load_namespaces=True) as io:
```

File: src/pynwb/__init__.py [hunk 2]
```diff
@@ -250,7 +250,8 @@ def can_read(path: str):
             {'name': 'extensions', 'type': (str, TypeMap, list),
              'doc': 'a path to a namespace, a TypeMap, or a list consisting paths to namespaces and TypeMaps',
              'default': None},
-            {'name': 'file', 'type': [h5py.File, 'S3File'], 'doc': 'a pre-existing h5py.File object', 'default': None},
+            {'name': 'file', 'type': [h5py.File, 'S3File', "RemFile"], 'doc': 'a pre-existing h5py.File object',
```
Comment (bendichter) [inline src/pynwb/__init__.py]: I actually think we might not need this change. I think we can just have `NWBHDF5IO` accept an `h5py.File`. Maybe I thought we needed a change due to a mistake I made mixing up the `RemFile.File` object and the `h5py.File` object.

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,7 @@
 ### Enhancements and minor changes
 - Add `NWBHDF5IO.can_read()`. @bendichter [#1703](https://github.com/NeurodataWithoutBorders/pynwb/pull/1703)
 - Add `pynwb.get_nwbfile_version()`. @bendichter [#1703](https://github.com/NeurodataWithoutBorders/pynwb/pull/1703)
+- Add support for `RemFile` in `NWBHDF5IO`. @bendichter [#1761](https://github.com/NeurodataWithoutBorders/pynwb/pull/1761)
```
Comment (bendichter) [inline CHANGELOG.md]:
```suggestion
```

File: src/pynwb/__init__.py [hunk 4]
```diff
@@ -250,7 +250,8 @@ def can_read(path: str):
             {'name': 'extensions', 'type': (str, TypeMap, list),
              'doc': 'a path to a namespace, a TypeMap, or a list consisting paths to namespaces and TypeMaps',
              'default': None},
-            {'name': 'file', 'type': [h5py.File, 'S3File'], 'doc': 'a pre-existing h5py.File object', 'default': None},
+            {'name': 'file', 'type': [h5py.File, 'S3File'], 'doc': 'a pre-existing h5py.File object',
+             'default': None},
```
Comment (bendichter) [inline src/pynwb/__init__.py]:
```suggestion
            {'name': 'file', 'type': [h5py.File, 'S3File'], 'doc': 'a pre-existing h5py.File object', 'default': None},
```

**Comment by bendichter** [comment]
@rly ok, ready for review

**Comment by rly** [comment]
Thanks. Could we make remfile Option 2, or even Option 1 on the streaming docs? See some related discussion https://github.com/NeurodataWithoutBorders/pynwb/issues/1791

File: docs/gallery/advanced_io/streaming.py [hunk 5]
```diff
@@ -122,6 +122,30 @@
 #        conda install -c conda-forge "h5py>=3.2"
 
 
+##################################################
+# Method 3: remfile
+# -----------------
+# remfile is another library that enables indexing and streaming of files in s3. remfile is simple, fast, and allows for
+# caching of data in the local filesystem. The one caveat of remfile is that it is a very new project that has not been
+# tested in a variety of use-cases. You can install remfile with pip:
```
Comment (rly) [inline docs/gallery/advanced_io/streaming.py]:
```suggestion
# caching of data in the local filesystem. The caveats of remfile are that it is a very new project that has not been
# tested in a variety of use-cases and caching options are limited compared to fsspec. You can install remfile with pip:
```

File: docs/gallery/advanced_io/streaming.py [hunk 6]
```diff
@@ -122,6 +122,30 @@
...
+#
+# .. code-block:: bash
+#
+#   pip install remfile
+#
+# Then remfile can be used similarly to fsspec
```
Comment (rly) [inline docs/gallery/advanced_io/streaming.py]:
```suggestion
```

File: environment-ros3.yml [hunk 7]
```diff
@@ -16,3 +16,6 @@ dependencies:
   - fsspec==2023.6.0
   - requests==2.28.1
   - aiohttp==3.8.3
+  - pip
+  - pip:
+    - remfile==0.1.7
```
Comment (rly) [inline environment-ros3.yml]:
```suggestion
    - remfile==0.1.9
```

**Review by rly — APPROVED** [review]

**Comment by magland** [comment]
Thanks @rly !

Here's a very rough benchmark I put together to compare timings of fsspec and remfile

https://github.com/scratchrealm/pynwb_streaming_benchmark

Pasting a snapshot of the readme:

The script main.py was run repeatedly in two settings: On dandihub and on a laptop in a home network.
The results are in results_dandihub.txt and results_home_network.txt, respectively.

Below is a summary of the average timings. My assessment is that remfile appears to be faster for initial load time while fsspec method appears to be faster for reading a 30 second sample of ephys data. On the home network, the fsspec method appears to be substantially slower than the remfile method for the initial load time.

Some limitations:
* Only applied to a single NWB file
* Only ran a limited number of trials
* Only ran on a single home network
* Didn't test ros3 method yet

---
**On dandihub**

Average Initial Load Time:

- fsspec: 7.74 seconds
- remfile: 6.14 seconds

Average 30sec Sample Read Time:
- fsspec: 7.90 seconds
- remfile: 8.20 seconds ​


---

**On Jeremy's Home Network**

Average Initial Load Time:
- fsspec: 37.23 seconds
- remfile: 9.42 seconds

Average 30sec Sample Read Time:
- fsspec: 47.61 seconds
- remfile: 56.95 seconds ​

**Pull request merged**